prow/config/inrepoconfig_test.go

```diff
@@ -631,3 +639,80 @@ func testDefaultProwYAMLGetter_RejectsNonGitHubRepo(clients localgit.Clients, t
 		t.Errorf("Error %v does not have expected message %s", err, expectedErrMsg)
 	}
 }
+
+type testClientFactory struct {
+	git.ClientFactory // This will be nil during testing, we override the functions that are used.
+	clientsCreated    int
+}
+
+func (cf *testClientFactory) ClientFor(org, repo string) (git.RepoClient, error) {
+	cf.clientsCreated++
+	// Returning this RepoClient ensures that only Fetch() is called and that Close() is not.
+	return &fetchOnlyNoCleanRepoClient{}, nil
+}
+
+type fetchOnlyNoCleanRepoClient struct {
+	git.RepoClient // This will be nil during testing, we override the functions that are allowed to be used.
+}
+
+func (rc *fetchOnlyNoCleanRepoClient) Fetch() error {
+	return nil
+}
+
+// TestInRepoConfigGitCacheConcurrency validates the following properties of InRepoConfigGitCache
+// - RepoClients are protected from concurrent use.
+// - Use of a RepoClient for one repo does not prevent concurrent use of a RepoClient for another repo.
+// - RepoClients are reused, only 1 client is created per repo.
+func TestInRepoConfigGitCacheConcurrency(t *testing.T) {
+	t.Parallel()
+
+	cf := &testClientFactory{}
+	cache := NewInRepoConfigGitCache(cf)
+	org, repo1, repo2 := "org", "repo1", "repo2"
+	// block channels are populated from the main thread, signal channels are read from the main thread.
+	signal := make(chan bool)
+	sharedRepo1State := 0
+
+	// Thread 1: gets a client for repo1, signals on signal1, then blocks on block1 before Clean()ing the repo1 client.
+	go func() {
+		client, err := cache.ClientFor(org, repo1)
+		if err != nil {
+			t.Errorf("Unexpected error getting repo client for thread 1: %v.", err)
+			return
+		}
+		sharedRepo1State++
+		client.Clean()
+		signal <- true
+	}()
+
+	// Thread 2: gets a client for repo1, signals success on signal2, then Clean()s the repo1 client.
+	go func() {
+		client, err := cache.ClientFor(org, repo1)
+		if err != nil {
+			t.Errorf("Unexpected error getting repo client for thread 2: %v.", err)
+			return
+		}
+		sharedRepo1State++
+		client.Clean()
+		signal <- true
+	}()
+
+	// Thread 3: gets a client for repo2, signals success on signal3, then Clean()s the repo2 client.
+	go func() {
+		client, err := cache.ClientFor(org, repo2)
+		if err != nil {
+			t.Errorf("Unexpected error getting repo client for thread 3: %v.", err)
+			return
+		}
+		client.Clean()
+		signal <- true
+	}()
+
+	for i := 0; i < 3; i++ {
```

Comment:
Question: I guess we're just testing that all 3 goroutines are able to finish? (And, if we don't get any of the signals that Golang will detect a deadlock?)

Reply:
We are waiting, otherwise the test might finish when the routines still run